## Mercedes-Benz Greener Manufacturing 

# Import Libraries and Datasets

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
train_data = pd.read_csv('C:/Users/sriji/MachineLearning/Final Project/Mercedes/Dataset/train/train.csv')

test_data = pd.read_csv('C:/Users/sriji/MachineLearning/Final Project/Mercedes/Dataset/test/test.csv')

--------------------------------------

# EDA

In [3]:
train_data.shape

(4209, 378)

In [4]:
test_data.shape

(4209, 377)

In [5]:
print(train_data.columns)

Index(['ID', 'y', 'X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8',
       ...
       'X375', 'X376', 'X377', 'X378', 'X379', 'X380', 'X382', 'X383', 'X384',
       'X385'],
      dtype='object', length=378)


In [6]:
print(test_data.columns)

Index(['ID', 'X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8', 'X10',
       ...
       'X375', 'X376', 'X377', 'X378', 'X379', 'X380', 'X382', 'X383', 'X384',
       'X385'],
      dtype='object', length=377)


In [7]:
train_data.head()

ID       y  X0 X1  X2 X3 X4 X5 X6 X8  ...  X375  X376  X377  X378  X379  \
0   0  130.81   k  v  at  a  d  u  j  o  ...     0     0     1     0     0   
1   6   88.53   k  t  av  e  d  y  l  o  ...     1     0     0     0     0   
2   7   76.26  az  w   n  c  d  x  j  x  ...     0     0     0     0     0   
3   9   80.62  az  t   n  f  d  x  l  e  ...     0     0     0     0     0   
4  13   78.02  az  v   n  f  d  h  d  n  ...     0     0     0     0     0   

   X380  X382  X383  X384  X385  
0     0     0     0     0     0  
1     0     0     0     0     0  
2     0     1     0     0     0  
3     0     0     0     0     0  
4     0     0     0     0     0  

[5 rows x 378 columns]

In [8]:
test_data.head()

ID  X0 X1  X2 X3 X4 X5 X6 X8  X10  ...  X375  X376  X377  X378  X379  X380  \
0   1  az  v   n  f  d  t  a  w    0  ...     0     0     0     1     0     0   
1   2   t  b  ai  a  d  b  g  y    0  ...     0     0     1     0     0     0   
2   3  az  v  as  f  d  a  j  j    0  ...     0     0     0     1     0     0   
3   4  az  l   n  f  d  z  l  n    0  ...     0     0     0     1     0     0   
4   5   w  s  as  c  d  y  i  m    0  ...     1     0     0     0     0     0   

   X382  X383  X384  X385  
0     0     0     0     0  
1     0     0     0     0  
2     0     0     0     0  
3     0     0     0     0  
4     0     0     0     0  

[5 rows x 377 columns]

In [9]:
train_data.describe()

ID            y          X10     X11          X12  \
count  4209.000000  4209.000000  4209.000000  4209.0  4209.000000   
mean   4205.960798   100.669318     0.013305     0.0     0.075077   
std    2437.608688    12.679381     0.114590     0.0     0.263547   
min       0.000000    72.110000     0.000000     0.0     0.000000   
25%    2095.000000    90.820000     0.000000     0.0     0.000000   
50%    4220.000000    99.150000     0.000000     0.0     0.000000   
75%    6314.000000   109.010000     0.000000     0.0     0.000000   
max    8417.000000   265.320000     1.000000     0.0     1.000000   

               X13          X14          X15          X16          X17  ...  \
count  4209.000000  4209.000000  4209.000000  4209.000000  4209.000000  ...   
mean      0.057971     0.428130     0.000475     0.002613     0.007603  ...   
std       0.233716     0.494867     0.021796     0.051061     0.086872  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
50%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
75%       0.000000     1.000000     0.000000     0.000000     0.000000  ...   
max       1.000000     1.000000     1.000000     1.000000     1.000000  ...   

              X375         X376         X377         X378         X379  \
count  4209.000000  4209.000000  4209.000000  4209.000000  4209.000000   
mean      0.318841     0.057258     0.314802     0.020670     0.009503   
std       0.466082     0.232363     0.464492     0.142294     0.097033   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       1.000000     0.000000     1.000000     0.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

              X380         X382         X383         X384         X385  
count  4209.000000  4209.000000  4209.000000  4209.000000  4209.000000  
mean      0.008078     0.007603     0.001663     0.000475     0.001426  
std       0.089524     0.086872     0.040752     0.021796     0.037734  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     0.000000     0.000000  
75%       0.000000     0.000000     0.000000     0.000000     0.000000  
max       1.000000     1.000000     1.000000     1.000000     1.000000  

[8 rows x 370 columns]

From the Analysis we understand that,

The Train dataset has 376 input features - which are classified either as 0 or 1.
The Train dataset has 1 output label 'y'- which is minutes spend on Test Bed.
The column ID is just a unique value for each record.

The Test dataset is similar to the Train data set, with the only difference that it does not has an output feature/column 

There are no null values in both train and test dataset

------------------

# If for any column(s), the variance is equal to zero, then you need to remove those variable(s)

In [10]:
col_with_zero_var = train_data.var()[train_data.var() == 0].index.values

In [11]:
print(col_with_zero_var)

['X11' 'X93' 'X107' 'X233' 'X235' 'X268' 'X289' 'X290' 'X293' 'X297'
 'X330' 'X347']


The above mentioned are the list of columns with variance equal to zero

It is verified in the next step

In [12]:
train_data[col_with_zero_var].var()

X11     0.0
X93     0.0
X107    0.0
X233    0.0
X235    0.0
X268    0.0
X289    0.0
X290    0.0
X293    0.0
X297    0.0
X330    0.0
X347    0.0
dtype: float64

In [13]:
train_df = train_data.drop(col_with_zero_var, axis=1)

In [14]:
test_df = test_data.drop(col_with_zero_var, axis=1)

The Columns with zero variance is removed in both train and test dataset

--------------------------------------

# Check for null and unique values for test and train sets.

In [15]:
cols_with_null = train_data.isna().sum()[train_data.isna().sum() > 0].index.values

len(cols_with_null)

0

In [16]:
cols_with_null = test_data.isna().sum()[test_data.isna().sum() > 0].index.values

len(cols_with_null)

0

In [17]:
np.sum(train_data.isna().sum())

#easier way

0

In [18]:
np.sum(test_data.isna().sum())

0

In [19]:
print(train_df.shape)
print(test_df.shape)

(4209, 366)
(4209, 365)


In [20]:
train_df = train_df.drop(['ID'],axis=1)
test_df = test_df.drop(['ID'],axis=1)

# ID Column does not contribute to prediction, hence it is dropped

In [21]:
print(train_df.shape)
print(test_df.shape)

(4209, 365)
(4209, 364)


-------------------------------------

# Apply label encoder

In [22]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4209 entries, 0 to 4208
Columns: 365 entries, y to X385
dtypes: float64(1), int64(356), object(8)
memory usage: 11.7+ MB


In [23]:
label_columns = train_df.describe(include=['object']).columns.values

label_columns

array(['X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8'], dtype=object)

In [24]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

In [25]:
for column in label_columns:
    le.fit( train_df[column].append(test_df[column]).values )
    
    train_df[column] = le.transform(train_df[column])
    
    test_df[column] = le.transform(test_df[column])

In [26]:
train_df.head()

y  X0  X1  X2  X3  X4  X5  X6  X8  X10  ...  X375  X376  X377  X378  \
0  130.81  37  23  20   0   3  27   9  14    0  ...     0     0     1     0   
1   88.53  37  21  22   4   3  31  11  14    0  ...     1     0     0     0   
2   76.26  24  24  38   2   3  30   9  23    0  ...     0     0     0     0   
3   80.62  24  21  38   5   3  30  11   4    0  ...     0     0     0     0   
4   78.02  24  23  38   5   3  14   3  13    0  ...     0     0     0     0   

   X379  X380  X382  X383  X384  X385  
0     0     0     0     0     0     0  
1     0     0     0     0     0     0  
2     0     0     1     0     0     0  
3     0     0     0     0     0     0  
4     0     0     0     0     0     0  

[5 rows x 365 columns]

In [27]:
test_df.head()

X0  X1  X2  X3  X4  X5  X6  X8  X10  X12  ...  X375  X376  X377  X378  \
0  24  23  38   5   3  26   0  22    0    0  ...     0     0     0     1   
1  46   3   9   0   3   9   6  24    0    0  ...     0     0     1     0   
2  24  23  19   5   3   0   9   9    0    0  ...     0     0     0     1   
3  24  13  38   5   3  32  11  13    0    0  ...     0     0     0     1   
4  49  20  19   2   3  31   8  12    0    0  ...     1     0     0     0   

   X379  X380  X382  X383  X384  X385  
0     0     0     0     0     0     0  
1     0     0     0     0     0     0  
2     0     0     0     0     0     0  
3     0     0     0     0     0     0  
4     0     0     0     0     0     0  

[5 rows x 364 columns]

--------------------------------

# Perform dimensionality reduction

In [28]:
from sklearn.decomposition import PCA

In [29]:
pca = PCA(n_components=0.98, svd_solver='full')

In [30]:
from sklearn.model_selection import train_test_split as tts

In [31]:
X = train_df.drop('y',axis=1)
y = train_df[['y']]

In [32]:
X.head()

X0  X1  X2  X3  X4  X5  X6  X8  X10  X12  ...  X375  X376  X377  X378  \
0  37  23  20   0   3  27   9  14    0    0  ...     0     0     1     0   
1  37  21  22   4   3  31  11  14    0    0  ...     1     0     0     0   
2  24  24  38   2   3  30   9  23    0    0  ...     0     0     0     0   
3  24  21  38   5   3  30  11   4    0    0  ...     0     0     0     0   
4  24  23  38   5   3  14   3  13    0    0  ...     0     0     0     0   

   X379  X380  X382  X383  X384  X385  
0     0     0     0     0     0     0  
1     0     0     0     0     0     0  
2     0     0     1     0     0     0  
3     0     0     0     0     0     0  
4     0     0     0     0     0     0  

[5 rows x 364 columns]

In [33]:
y.head()

y
0  130.81
1   88.53
2   76.26
3   80.62
4   78.02

In [34]:
X_train,X_val,y_train,y_val = tts(X,y,test_size=0.2)

In [35]:
print(X_train.shape)
print(X_val.shape)
print(y_train.shape)
print(y_val.shape)

(3367, 364)
(842, 364)
(3367, 1)
(842, 1)


In [36]:
pca.fit(X_train)

PCA(copy=True, iterated_power='auto', n_components=0.98, random_state=None,
    svd_solver='full', tol=0.0, whiten=False)

In [37]:
X_train_transformed = pca.transform(X_train)
X_val_transformed = pca.transform(X_val)

In [38]:
print(X_train_transformed.shape)
print(X_val_transformed.shape)

(3367, 12)
(842, 12)


In [39]:
pca.n_components_

12

No of features is reduced from 364 to 12 with a variance of 98%

In [40]:
pca.explained_variance_ratio_

#column-wise ratio

array([0.4074617 , 0.2167973 , 0.13069898, 0.10937377, 0.08257013,
       0.01403983, 0.00659401, 0.00390544, 0.00258134, 0.00211402,
       0.00206526, 0.00182526])

In [41]:
np.sum(pca.explained_variance_ratio_)

#sum of all columns ratio -> equals to the mentioned n_components/variance ration mentioned (0.98)

0.9800270620668659

In [42]:
pca_X_train = pd.DataFrame(X_train_transformed)
pca_X_val = pd.DataFrame(X_val_transformed)

In [43]:
pca_X_train.head()

0          1          2          3          4         5         6  \
0  21.561872  -6.782742   9.338388 -16.318249  -7.406103 -2.155935  1.606314   
1  24.180209 -14.482083  10.611954   3.761914 -11.795826 -0.105138  1.210185   
2  12.919232  -3.043445  -5.981902  13.048398  -4.477671 -2.732915  1.482090   
3  23.745240  17.840101  12.600751  -9.129929  -7.832127 -1.876681  3.571009   
4  -9.084895  19.476882  -6.471583  -5.585605  -5.022400 -3.452219  0.944922   

          7         8         9        10        11  
0 -1.067201 -1.378841 -1.528457 -0.545948 -0.684069  
1  0.283719  0.381616  1.366942 -0.585807 -0.142627  
2  0.021365  0.135561 -1.251054 -0.506489  3.206964  
3  1.698630 -2.278660 -1.127000  4.242075  0.910546  
4  1.023003 -0.332420  0.045746 -1.789288 -1.781822

In [44]:
pca_test = pd.DataFrame(pca.transform(test_df))

In [45]:
pca_test.shape

(4209, 12)

In [46]:
pca_test.head()

0          1          2          3          4         5         6  \
0  16.071499  13.636313  14.130031   6.476482  10.205897  6.760611 -2.003966   
1 -15.523569  -8.782973  -7.533840  -2.224322  12.188544  0.940021  3.825748   
2  12.524333  -2.847423 -12.228334  13.007589  -3.172858 -2.604299 -1.095883   
3  14.015164  14.167991  16.633873  -5.641396   2.274414 -3.892969 -2.851942   
4 -12.361657   1.037994  18.222906   6.900893  -1.399031 -1.425170  0.975054   

          7         8         9        10        11  
0  2.296091  1.084141 -2.183080 -0.555093  3.422961  
1  2.060704 -0.463736  0.308401  0.814781 -0.069073  
2  0.455374  0.309586 -1.287259  0.104537  3.249042  
3  1.984179 -0.373864 -2.433445 -1.001940  3.651863  
4 -3.066153 -1.395759  0.374615 -0.001899  0.203369

---------------------------------------------------

# Predict your test_df values using XGBoost

In [47]:
from xgboost import XGBRegressor

In [48]:
model = XGBRegressor(objective='reg:linear',learning_rate=0.1)

In [49]:
model.fit(pca_X_train,y_train)

[04:43:41] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.0.0/src/objective/regression_obj.cu:167: reg:linear is now deprecated in favor of reg:squarederror.


XGBRegressor(base_score=0.5, booster=None, colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
             importance_type='gain', interaction_constraints=None,
             learning_rate=0.1, max_delta_step=0, max_depth=6,
             min_child_weight=1, missing=nan, monotone_constraints=None,
             n_estimators=100, n_jobs=0, num_parallel_tree=1,
             objective='reg:linear', random_state=0, reg_alpha=0, reg_lambda=1,
             scale_pos_weight=1, subsample=1, tree_method=None,
             validate_parameters=False, verbosity=None)

In [50]:
pred_y_val = model.predict(pca_X_val)

In [51]:
print(pred_y_val.shape)
print(y_val.shape)

(842,)
(842, 1)


In [52]:
y_val

y
3447  104.18
3824  108.40
1979  108.99
1542  108.42
1752   89.75
...      ...
3155   96.74
1029   89.65
1803   72.94
3265   99.95
916    89.96

[842 rows x 1 columns]

In [63]:
from sklearn.metrics import mean_squared_error as mserror
from sklearn.metrics import accuracy_score as acc

In [64]:
mse = mserror(pred_y_val,y_val)
print(mse)

83.31789593973396


In [65]:
print(np.sqrt(mse))

9.127863711719954


RMSE of XGB Regressor algorithm is 9.13

------------------------------

# Building Random Forest Model to compare the RMSE Scores

In [56]:
from sklearn.ensemble import RandomForestRegressor

In [58]:
regr = RandomForestRegressor(max_depth=2,random_state=0)

In [59]:
regr.fit(pca_X_train,y_train)

D:\Software-Installation\Anaconda\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
D:\Software-Installation\Anaconda\lib\site-packages\ipykernel_launcher.py:1: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  """Entry point for launching an IPython kernel.


RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=2,
                      max_features='auto', max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, n_estimators=10,
                      n_jobs=None, oob_score=False, random_state=0, verbose=0,
                      warm_start=False)

In [61]:
y_val_pred = regr.predict(pca_X_val)

In [68]:
mse = mserror(y_val,y_val_pred)
print(mse)

117.30385611954712


In [69]:
print(np.sqrt(mse))

10.830690472889858


RMSE of Random Forest Regressor is 10.84

## Hence XGB Regressor is better suited for this problem

In [76]:
y_test_pred = model.predict(pca_test)

In [77]:
y_test = pd.DataFrame(y_test_pred)
y_test.head()

0
0   77.847382
1   95.422340
2   78.487282
3   77.193657
4  124.727554

## Hence the values are predicted for TEST dataset using the XGB Regressor Algorithm

------------------------------------------------------------------------------